### PR TITLE
Add new Haskell linter

### PR DESCRIPTION
**Pull request by jez**
This PR adds a new Haskell linter (based on GHC). As such, it can take advantage
of the pre-existing `HandleGHCFormat` function. Also, we update the docs.

- **Add stack-build linter for Haskell**

  The stack-build linter works better than the other two linters when you're
  working with an entire Haskell project. It builds the project entirely and
  reports any errors.

  The other two Haskell GHC linters only work on single files, which can
  result in spurious errors (for example, not being able to find imports).

- **Document all available Haskell linters**

### CODE REVIEW

**Comment by jez** [comment]
After looking at it a little longer, I realized that the `stack-ghc-mod` linter is actually more in line with what I was going for:

- uses `stack` to correctly resolve project dependencies
- just *checks* the file, instead of actually building the whole thing like `stack-build` does.

That being said, I think it's valuable to have both. `stack-ghc-mod` requires an extra dependency (`ghc-mod`). `stack-build` has the nice side effect of also building your project (you can imagine that people might want to build their project on save). And above all, keeping `stack-build` has minimal maintenance cost, because it re-uses `HandleGHCFormat`.

TL;DR: `stack-ghc-mod` is more of what I wanted, but `stack-build` is still fine to add the new `stack-build` linter.

File: ale_linters/haskell/ghc.vim [hunk 1]
```diff
@@ -16,3 +16,12 @@ call ale#linter#Define('haskell', {
 \   'command': 'stack ghc -- -fno-code -v0 %t',
 \   'callback': 'ale#handlers#haskell#HandleGHCFormat',
 \})
+
+call ale#linter#Define('haskell', {
```
Comment (w0rp) [inline ale_linters/haskell/ghc.vim]: Could you move this to a separate file? The other two should probably also be in their own files.

**Comment by jez** [comment]
> Will this generate binary files? That may surprise and annoy people.

Yes, this will. My reasoning why it wasn't a huge deal was that

- Stack's "big idea" is reproducible builds. Running `stack build` is as close to idempotent as you can get.
- ALE is async. Even if the code generation takes a long time, at least it's not blocking the UI.
- `stack-build` provides a nice alternative to `stack-ghc` and `stack-ghc-mod` in that it requires no external dev tools and works with an entire project.

Unfortunately, there's no support with stack to just typecheck: commercialhaskell/stack#977.

What do you think? Is it annoying enough that we shouldn't include it? Could we make it opt-in as opposed to on by default?

File: ale_linters/haskell/ghc.vim (hunk 1 above)
Comment (jez) [inline ale_linters/haskell/ghc.vim]: Done.

**Comment by jez** [comment]
Should be ready for round 2, thanks!

**Comment by w0rp** [comment]
I will summon Haskell users who have commented before.

@wizzup @azure-satellite @parsonsmatt @jwoudenberg @WhiteAbeLincoln @habibalamin Does anyone mind if a Haskell linter generates binary files?

**Comment by habibalamin** [comment]
As long as it's stored in an OS-provided tempfolder, cleaned up ASAP, and is space-efficient (depending on the program being linted), I wouldn't mind too much. However, I also don't know if I *should* mind.

However, I'm not sure how a project-level linter would work when embedded in a text editor. Is it necessary to have? What can it do that a file-level linter can't?

**Comment by habibalamin** [comment]
Actually, forget about storing it in a tempfolder. If it just builds the project in the same location `stack build` normally does, it shouldn't be an issue. I just don't want any stray copies of the built project laying around, but if it's building where I'm building, I'm okay with it.

**Comment by w0rp** [comment]
If it builds into the regular build directory, which it should do, then it'll probably be fine.

**Comment by habibalamin** [comment]
I don't have any objections, but someone more knowledgeable than me may is where I'll leave it. I'm happy with this PR being merged.

**Comment by w0rp** [comment]
I'll merge this, and let people try it.

**Comment by w0rp** [comment]
Cheers! :beers:

**Comment by parsonsmatt** [comment]
This looks great to me! Running `build` is pretty typical. The only suggestion I would make is using `--fast` to build without optimizations, which is about twice as fast on my work project.

**Comment by w0rp** [comment]
@jez Do you want to try using `--fast`? Sounds like a good idea.

**Comment by jez** [comment]
Good idea. Added in #754.
